### PR TITLE
Clean up shadow code & cast shadows from Directional Lights

### PR DESCRIPTION
**Pull request by fredroy**

### CODE REVIEW

File: modules/SofaOpenglVisual/Light.cpp [hunk 1]
```diff
@@ -348,6 +394,196 @@ void DirectionalLight::draw(const core::visual::VisualParams* )
...
+
+    float zNear, left, bottom;
+    float zFar, right, top;
+
+    zNear, left, bottom = -1e10;
+    zFar = right = top = 1e10;
+
+    Light::preDrawShadow(vp);
+    const Vector3& dir = d_direction.getValue();
+    //TODO
```
Comment (damienmarchal) [inline modules/SofaOpenglVisual/Light.cpp]: Can you give some details on what is the TODO for ?

File: modules/SofaOpenglVisual/Light.cpp [hunk 2]
```diff
@@ -348,6 +394,196 @@ void DirectionalLight::draw(const core::visual::VisualParams* )
...
+    glViewport(0, 0, m_shadowTexWidth, m_shadowTexHeight);
+
+    glClearColor(1.0, 1.0, 1.0, 1.0);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    glEnable(GL_DEPTH_TEST);
+}
+
+GLuint DirectionalLight::getDepthTexture()
+{
+    //return debugVisualShadowTexture;
```
Comment (damienmarchal) [inline modules/SofaOpenglVisual/Light.cpp:571]: Are these comment still needed ?

File: modules/SofaOpenglVisual/Light.cpp [hunk 3]
```diff
@@ -692,7 +919,32 @@ void SpotLight::computeOpenGLProjectionMatrix(GLfloat mat[16], float width, floa
     mat[7] = 0.0;
     mat[11] = -1.0;
     mat[15] = 0.0;
+
```
Comment (damienmarchal) [inline modules/SofaOpenglVisual/Light.cpp:928]: Do we need to keep the fllowing commented part ?

**Comment by damienmarchal** [comment]
Hi Fred,

Thank you for your pull-request. 
It looks really nice. 

Some user friendly stuff...
- I would be great to have more details in the help string of the added Data. Currently it is more or less duplicating the name; maybe you can say what is it used for in term of rendering and shadowing.
- I have now clear opinion on that but maybe we should also add the default values in the docstring (I found it very annoying for the users to look in the source code to know what the default values are). 
- In the same spirit I think that somewhere VMS should be documented to the user. Maybe in the component helpstring when it is registered ? 

I will test that by thursday unless someone else do that before.

**Comment by fredroy** [comment]
I added some informations and taken into account your remarks.
As for the default value into the docstring, I guess it should be done automatically by some mechanism (e.g in the GUI for example while displaying the docstring)

**Comment by damienmarchal** [comment]
Hi fred,

You are right, an automated system to report the default values is better.  

Your  "[shadowing] " in the docstring make me wonder a more general question about datas that only have a meaning when an other parameter is enable. 

To make a more user friendly sofa we should: 
    - test the data this kind of pattern in the init function:
```cpp
if( !d_shadowsEnabled.getValue() && d_shadowTextureSize.isSet()){
   msg_warning(this) << "The shadow are not enabled but the shadowTextureSize parameter is set.   "
                                            "To remove this warning you can either :   "
                                            "-  enable the 'shadow' parameter.  " 
                                            "-  remove the use of the shadowTextureSize parameter."
}

    - group and gray/ungrayed in the GUI the connected parameters. Do you think we could use the widget string in the Data field  for that ?
 
DM.

**Comment by fredroy** [comment]
- I agree that we should test data's value more thoroughly  ; the only problem is that there are potentially thousands of combination to test (depending the component obviously)
 - as for the data gathering in the GUI, I am not sure but it is already done for Data related to visualization (tab "Visualization" when you double-click the component). If I remember well, the process checks that the data's name contains "draw" or "show" words.  The problem is, how to tell the GUI that for Lights, it has to group all d_shadow* in a "Shadow" tab ? 
Otherwise, a solution could be to to add a "Collection of Data" type into SofaObject and register it at the construction. E.g 
```
//In Light
DataCollection dc_shadow("Shadowing");
this->addCollectionData(dc_shadow);
dc_shadow.add(d_shadowEnabled);
dc_shadow.add(d_shadowTextureSize);

//In the GUI
foreach(component->getDataCollections() as dc )
{
    tab = createNewTab(dc->getName())
    foreach(dc.getAllData as data )
        tab.add(data)
}
```

**Comment by damienmarchal** [comment]
In the case of the Light (and actually this is true to most of the sofa component) I see several very simple tests to add to validate the inputes. But you are totally right this cannot be done automatically (unless with a kind of 
code-path/coverage analysis) so the developper have to write at least the easy one.

**Comment by damienmarchal** [comment]
I tested the component and the examples and this sound fine, 
I merge it so that I can work on it to add more checking on its inputs.

**Comment by matthieu-nesme** [comment]
You just merged a branch that adds tons of warnings... Be careful please.

**Comment by matthieu-nesme** [comment]
including real problems...

**Comment by damienmarchal** [comment]
Yes I just noticed that.

**Comment by matthieu-nesme** [comment]
#86

**Pull request merged**